Haskell Parser Combinators/article.md

```diff
@@ -0,0 +1,1375 @@
+4. The backtracking mechanism may, in turn, lead to exponential time complexities.
+
+This shouldn't, however, discourage you from their use, as their benefits outweigh the drawbacks. Ultimately, the parsing library of your choice should be chosen based on your needs, and parser combinators are a good fit for most use cases.
+
+## Implementing a simple parser combinator from scratch
+
+<!--
+* make an analogy: there is a tape, and we back move forward (parsed) or
+  backward (failed) on this tape.
+* datatype to parse String
+* satisfy and instances for Monad and Alternative
+* maybe explain what Alternative is, since beginners may not be familiar with it
+-->
+
+Before we delve into `megaparsec`, it may be interesting to implement our parser combinator from scratch. We will make a simple one that should hopefully help you understand how parser combinators work.
+
+If you instead prefer to jump right into the action, then skip to the [Megaparsec tutorial](#megaparsec-tutorial).
+
+### Implementation
+
+We will begin by importing the appropriate definitions in a new file called `Parser.hs`:
+
+```hs
+module Parser where
+
+import Control.Applicative (Alternative (..))
+import Data.List (nub)
+```
+
+We will create two data structures that represent our parser.
+
+The first of them represents some error that may happen while parsing. For now, let's have three errors: one that is raised when we are expecting input but there is nothing to consume (`EndOfInput`), one when we find an unexpected character (`Unexpected`), and one for any error messages that the user might want to raise (`CustomError`).
+
+Our second structure is the actual parser. It's represented as a function that takes some input and either returns a list of errors, if `Left`, or the result of parsing the input together with the rest of the input (that was not parsed), if `Right`.
+
+```hs
+data Error i e
+  = EndOfInput  -- Expected more input, but there is nothing
+  | Unexpected i  -- We didn't expect to find this element
+  | CustomError e  -- Extra errors the user may want to create
+  | Empty  -- Used in `Alternative` implementation of `empty`
+  deriving (Eq, Show)
+
+newtype Parser i e a = Parser
+  { runParser :: [i] -> Either [Error i e] (a, [i])
+  }
+```
+
+This is the meaning of each of the type variables:
+* `i`: The input stream. For most cases, we will use `Char`, as `[Char]` is the same as `String`. It's interesting to notice that keeping this as a type variable allows us to parse things that are not strings as well.
+* `e`: The type of custom error messages. If we don't have those, we may use `Void` instead.
+* `a`: The result of our parsing function. It represents the structure parsed from our consumed input.
+
+Let's begin by creating the most primitive function: `satisfy`. It takes some predicate that operates on the current character, and advances the parser, if possible, returning the consumed character.
```

Comment:
Is it possible to have a bit simpler structure here?

takes a predicate that operates on the current character, advances the parser (if possible), and returns the consumed character.

Reply:
I reworded a bit, maybe it's easier to understand now.